In [1]:
# uncomment this command if you get an error from the following cell saying that 'highs' solver is not recognized
# !pip install --upgrade --user scipy


In [2]:
import numpy as np
from scipy.optimize import linprog

# Define cost per serving (Chicken, Rice, Broccoli)
cost = [2.50, 0.80, 1.50]  # Minimize: 2.5x + 0.8y + 1.5z

# Define nutrition constraints matrix (Calories, Protein, Fat)
# Each row corresponds to a nutrient requirement
# Each column represents a food item (Chicken, Rice, Broccoli)
nutrition = [
    [250, 130, 50],   # Calories per serving
    [30, 3, 4],       # Protein per serving
    [8, 0.5, 0.2]     # Fat per serving
]

# Define minimum daily intake for each nutrient
min_nutrition = [2000, 50, 10]  # [Calories, Protein, Fat]

# Convert constraints to standard form (Ax ≤ b)
A_ub = -1 * np.array(nutrition)  # Flip sign to convert ≥ to ≤
b_ub = -1 * np.array(min_nutrition)

# Define bounds for each food item (no negative servings)
bounds = [(0, None), (0, None), (0, None)]

# Solve the linear programming problem
result = linprog(c=cost, A_ub=A_ub, b_ub=b_ub, bounds=bounds, method="highs")

# Print results
if result.success:
    servings = result.x
    print("Optimal daily servings:")
    print(f"  Chicken: {servings[0]:.2f} servings")
    print(f"  Rice: {servings[1]:.2f} servings")
    print(f"  Broccoli: {servings[2]:.2f} servings")
    print(f"Total cost: ${result.fun:.2f}")
else:
    print("Optimization failed:", result.message)

# TODO: Extend the problem by adding a constraint for Vitamin C intake.
# Steps to update the code:



Optimal daily servings:
  Chicken: 0.33 servings
  Rice: 14.75 servings
  Broccoli: 0.00 servings
Total cost: $12.62


In [3]:
# Write the answers to the questions in this cell
#2. Update the `nutrition` matrix by adding a new row for Vitamin C:
#    nutrition.append([0, 0, 50])  # New row for Vitamin C intake
nutrition.append([0, 0, 50])  # New row for Vitamin C intake


In [4]:
# 3. Update the `min_nutrition` list to include the new requirement:
#    min_nutrition.append(30)  # At least 30 mg of Vitamin C required per day

min_nutrition.append(30)  # At least 30 mg of Vitamin C required per day


In [5]:
#4.Re-run the linprog function and observe the new results.


import numpy as np
from scipy.optimize import linprog

# Define cost per serving (Chicken, Rice, Broccoli)
cost = [2.50, 0.80, 1.50]  # Minimize: 2.5x + 0.8y + 1.5z

# Define nutrition constraints matrix (Calories, Protein, Fat, Vitamin C)
nutrition = [
    [250, 130, 50],   # Calories per serving
    [30, 3, 4],       # Protein per serving
    [8, 0.5, 0.2],    # Fat per serving
    [0, 0, 50]        # Vitamin C per serving
]

# Define minimum daily intake for each nutrient (Calories, Protein, Fat, Vitamin C)
min_nutrition = [2000, 50, 10, 30]

# Convert constraints to standard form (Ax ≤ b)
A_ub = -1 * np.array(nutrition)  # Flip sign to convert ≥ to ≤
b_ub = -1 * np.array(min_nutrition)

# Define bounds for each food item (no negative servings)
bounds = [(0, None), (0, None), (0, None)]

# Solve the linear programming problem
result = linprog(c=cost, A_ub=A_ub, b_ub=b_ub, bounds=bounds, method="highs")

# Print results
if result.success:
    servings = result.x
    print("Optimal daily servings:")
    print(f"  Chicken: {servings[0]:.2f} servings")
    print(f"  Rice: {servings[1]:.2f} servings")
    print(f"  Broccoli: {servings[2]:.2f} servings")
    print(f"Total cost: ${result.fun:.2f}")
else:
    print("Optimization failed:", result.message)


Optimal daily servings:
  Chicken: 0.33 servings
  Rice: 14.52 servings
  Broccoli: 0.60 servings
Total cost: $13.34


### Reflection on Diet Optimization

#### 1. Before adding Vitamin C:
- The optimal daily servings were:
  - **Chicken:** 0.33 servings  
  - **Rice:** 14.75 servings  
  - **Broccoli:** 0.00 servings  
- **Total cost:** **$12.62**  

#### 2. After adding Vitamin C:
- The optimal daily servings changed to:
  - **Chicken:** 0.33 servings (same as before)  
  - **Rice:** 14.52 servings (slightly reduced)  
  - **Broccoli:** 0.60 servings (increased to meet Vitamin C requirement)  
- **Total cost:** **$13.34** (increased by **$0.72**)  

#### 3. Trade-offs in diet planning:
- Adding a **Vitamin C requirement** forced the model to **increase broccoli consumption**, as it was the only available source of Vitamin C.  
- This **slightly reduced rice servings** while keeping chicken unchanged.  
- The **cost increased** because broccoli is **more expensive per serving** than rice.  
- This highlights the **trade-off between cost and nutrition**—a balanced diet may require spending more to meet essential nutrient needs.  
